_posts/2021-04-02-being-productive-with-typescript.md

```diff
@@ -0,0 +1,477 @@
+   OPEN: '#2266FF',
+   CLOSED: '#888888',
+   'IN PROGRESS': '#FF8800'
+ }
+ return map[status];
+}
+```
+This is more than just adding Typescript, we refactored the [switch](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/switch){:target="_blank" rel="noopener noreferrer"} into a _lookup map_.
+For the [switch](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/switch){:target="_blank" rel="noopener noreferrer"} we had in our example, this can be done. But that is not always the case.
+The _lookup map_ has a type which is generated entirely within Typescript.
+[Record](https://www.typescriptlang.org/docs/handbook/utility-types.html#recordkeystype){:target="_blank" rel="noopener noreferrer"} is the utility type to be used, but in the comment below I wrote a more expressive way that achieves the same.
+So how does it help us? Reimagine that a new status is added. 
+I adapt the `agendaItemStatus`, hit ctrl-s (it starts a build).
+In a matter of seconds I will be informed that property 'DELETED' is missing on line _x_ of file _y_.
+Now we are using the potential of Typescript
+
+- Readability
+- Code hinting
+- Compile-time errors
+
+# Type aliases, DTO's and Utility Types
+
+```typescript
+interface CalenderItem {
+ duration?: number,
+ startTime: string | Date, // yyyy-MM-ddThh:mm:ss.sssZ
+ dueDate?: string | Date // yyyy-MM-dd
+}
+
+
+export function mapCalenderItem(dto: CalenderItem) {
+ return {
+   ...dto,
+   startTime: new Date(dto.startTime),
+   dueDate: dto.dueDate ? new Date(dto.dueDate) : undefined
+ }
+}
+
+
+export function remapCalenderItem(item: CalenderItem) {
+ return {
+   ...item,
+   startTime: (item.startTime as Date).toISOString(),
+   dueDate: item.dueDate 
+     ? (item.dueDate as Date).toISOString() 
+     : undefined
+ }
+}
+```
+The above code might need some explanation (so much for self-explanatory). 
+My datasource is a REST API, therefore it communicates dates as string.
+But as a developer you do not want to pass strings around and convert them every time you need them to be a Date.
+So as soon as I receive the dates, I change them into Date objects (`mapCalenderItem`). 
+And when I post them back to backend, I parse them back to strings (`remapCalenderItem`).
+
+The first thing to notice here, is that the interface is unstable. 
+`startTime` and `dueDate` are using a [union type](https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html#union-types){:target="_blank" rel="noopener noreferrer"}, this means it can be either a string or a Date.
+Then there is also the difference for Date and Time, having a ymd-string or a isoString respectively.
+Finally there is the [type assertion](https://www.typescriptlang.org/docs/handbook/basic-types.html#type-assertions){:target="_blank" rel="noopener noreferrer"}. 
+Type assertions happen when Typescript is getting in your way. It is adding negative value at this point.
```

Comment:
To me it wasn't clear what type assertion without clicking on that link, maybe it would be better to repeat the piece of code you're referring to?

Reply:
Added a explanation that refers back to the code.